In [4]:
#Importing libraries
from keras import models
from keras import layers
import numpy as np
import tensorflow
from keras.datasets import imdb
from tensorflow.keras.metrics import Precision


In [5]:
#Dividing Imdb dataset into train data and test data
(train_data, train_labels), (test_data,test_labels) = imdb.load_data(num_words = 10000)

In [6]:
#As the input data is in integers,we have to chage that to tensors of 1's and 0's.
def vectorize_sequences(sequences, dimension = 10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1
    return results

In [7]:
x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [8]:
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')

In [9]:
x_train.shape

(25000, 10000)

In [10]:
input_dimension = 10000

In [11]:
#Defining the model architecture
model = models.Sequential([
    layers.Dense(16, activation = 'relu', input_shape = (input_dimension,)),
    layers.Dense(16, activation = 'relu'),
    layers.Dense(1, activation = 'sigmoid')
])

In [12]:
#Defining the optimizer, loss function and metric
model.compile(optimizer = 'rmsprop',
            loss = 'binary_crossentropy',
            metrics = [Precision()])

In [13]:
#The train data is divided into validation set to guide the training process.
x_val = x_train[:10000]
p_x_train = x_train[10000:]
y_val = y_train[:10000]
p_y_train = y_train[10000:]

In [14]:
history = model.fit(p_x_train,
                    p_y_train,
                    epochs = 10,
                    batch_size = 512,
                    validation_data = (x_val,y_val))

Epoch 1/10
30/30 [==============================] - 2s 48ms/step - loss: 0.5367 - precision: 0.7933 - val_loss: 0.4177 - val_precision: 0.8836
Epoch 2/10
30/30 [==============================] - 1s 22ms/step - loss: 0.3407 - precision: 0.8865 - val_loss: 0.3239 - val_precision: 0.8596
Epoch 3/10
30/30 [==============================] - 1s 21ms/step - loss: 0.2543 - precision: 0.9099 - val_loss: 0.2876 - val_precision: 0.8787
Epoch 4/10
30/30 [==============================] - 1s 19ms/step - loss: 0.2069 - precision: 0.9257 - val_loss: 0.2973 - val_precision: 0.9251
Epoch 5/10
30/30 [==============================] - 1s 19ms/step - loss: 0.1723 - precision: 0.9406 - val_loss: 0.3021 - val_precision: 0.8355
Epoch 6/10
30/30 [==============================] - 1s 18ms/step - loss: 0.1500 - precision: 0.9497 - val_loss: 0.3132 - val_precision: 0.9263
Epoch 7/10
30/30 [==============================] - 1s 19ms/step - loss: 0.1290 - precision: 0.9612 - val_loss: 0.3213 - val_precision: 0.9216

In [15]:
results = model.evaluate(x_test,y_test)

782/782 [==============================] - 1s 2ms/step - loss: 0.3580 - precision: 0.8860


In [16]:
print(results)

[0.3579956293106079, 0.8859795928001404]


In [17]:
model.compile(optimizer = 'rmsprop',
            loss = 'binary_crossentropy',
            metrics = ['acc']) #Change in metric

history = model.fit(p_x_train,
                    p_y_train,
                    epochs = 4,
                    batch_size = 512,
                    validation_data = (x_val,y_val))

results = model.evaluate(x_test,y_test)

print(results)

Epoch 1/4
30/30 [==============================] - 2s 53ms/step - loss: 0.0822 - acc: 0.9755 - val_loss: 0.3518 - val_acc: 0.8737
Epoch 2/4
30/30 [==============================] - 1s 18ms/step - loss: 0.0635 - acc: 0.9857 - val_loss: 0.3664 - val_acc: 0.8798
Epoch 3/4
30/30 [==============================] - 1s 17ms/step - loss: 0.0562 - acc: 0.9864 - val_loss: 0.3982 - val_acc: 0.8770
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.4309 - acc: 0.8655
[0.4309108853340149, 0.8655200004577637]


In [18]:
model = models.Sequential([
    layers.Dense(16, activation = 'relu', input_shape = (input_dimension,)),
    layers.Dense(16, activation = 'relu'),
    layers.Dense(16, activation = 'relu'),# Added a thrid layer to check it's effect
    layers.Dense(1, activation = 'sigmoid')
])
model.compile(optimizer = 'rmsprop',
            loss = 'binary_crossentropy',
            metrics = ['acc']) 

history = model.fit(p_x_train,
                    p_y_train,
                    epochs = 4,
                    batch_size = 512,
                    validation_data = (x_val,y_val))

results = model.evaluate(x_test,y_test)

print(results)

Epoch 1/4
30/30 [==============================] - 3s 59ms/step - loss: 0.5471 - acc: 0.7839 - val_loss: 0.4211 - val_acc: 0.8583
Epoch 2/4
30/30 [==============================] - 1s 23ms/step - loss: 0.3385 - acc: 0.8884 - val_loss: 0.3290 - val_acc: 0.8700
Epoch 3/4
30/30 [==============================] - 1s 23ms/step - loss: 0.2468 - acc: 0.9137 - val_loss: 0.2960 - val_acc: 0.8813
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.3027 - acc: 0.8767
[0.3027161955833435, 0.8767200112342834]


In [19]:
model = models.Sequential([
    layers.Dense(16, activation = 'relu', input_shape = (input_dimension,)),
    layers.Dense(16, activation = 'relu'),
    layers.Dense(16, activation = 'relu'),# Added a thrid layer to check it's effect
    layers.Dense(16, activation = 'relu'),# Added a fourth layer
    layers.Dense(1, activation = 'sigmoid')
])
model.compile(optimizer = 'rmsprop',
            loss = 'binary_crossentropy',
            metrics = ['acc']) 

history = model.fit(p_x_train,
                    p_y_train,
                    epochs = 4,
                    batch_size = 512,
                    validation_data = (x_val,y_val))

results = model.evaluate(x_test,y_test)

print(results)

Epoch 1/4
30/30 [==============================] - 3s 59ms/step - loss: 0.6052 - acc: 0.6577 - val_loss: 0.5170 - val_acc: 0.8320
Epoch 2/4
30/30 [==============================] - 1s 20ms/step - loss: 0.4480 - acc: 0.8708 - val_loss: 0.4490 - val_acc: 0.8487
Epoch 3/4
30/30 [==============================] - 1s 21ms/step - loss: 0.3058 - acc: 0.9158 - val_loss: 0.3186 - val_acc: 0.8826
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.3112 - acc: 0.8766
[0.31116050481796265, 0.8765599727630615]


In [20]:
#Changed the hidden unit dimensions
model = models.Sequential([
    layers.Dense(64, activation = 'relu', input_shape = (input_dimension,)),
    layers.Dense(32, activation = 'relu'),
    layers.Dense(1, activation = 'sigmoid')
])
model.compile(optimizer = 'rmsprop',
            loss = 'binary_crossentropy',
            metrics = ['acc']) 

history = model.fit(p_x_train,
                    p_y_train,
                    epochs = 4,
                    batch_size = 512,
                    validation_data = (x_val,y_val))

results = model.evaluate(x_test,y_test)

print(results)

Epoch 1/4
30/30 [==============================] - 3s 59ms/step - loss: 0.5055 - acc: 0.7539 - val_loss: 0.4506 - val_acc: 0.7893
Epoch 2/4
30/30 [==============================] - 1s 29ms/step - loss: 0.2973 - acc: 0.8871 - val_loss: 0.2946 - val_acc: 0.8859
Epoch 3/4
30/30 [==============================] - 1s 28ms/step - loss: 0.2185 - acc: 0.9195 - val_loss: 0.2781 - val_acc: 0.8870
Epoch 4/4
782/782 [==============================] - 3s 3ms/step - loss: 0.3467 - acc: 0.8587
[0.34669992327690125, 0.858680009841919]


In [23]:
#Changed the hidden unit dimensions
model = models.Sequential([
    layers.Dense(16, activation = 'relu', input_shape = (input_dimension,)),
    layers.Dense(16, activation = 'relu'),
    layers.Dense(1, activation = 'sigmoid')
])
model.compile(optimizer = 'rmsprop',
            loss = 'binary_crossentropy',
            metrics = ['mse']) 

history = model.fit(p_x_train,
                    p_y_train,
                    epochs = 4,
                    batch_size = 512,
                    validation_data = (x_val,y_val))

results = model.evaluate(x_test,y_test)

print(results)

Epoch 1/4
30/30 [==============================] - 3s 76ms/step - loss: 0.5536 - mse: 0.1848 - val_loss: 0.4250 - val_mse: 0.1292
Epoch 2/4
30/30 [==============================] - 1s 27ms/step - loss: 0.3526 - mse: 0.1027 - val_loss: 0.3291 - val_mse: 0.0969
Epoch 3/4
30/30 [==============================] - 5s 156ms/step - loss: 0.2623 - mse: 0.0734 - val_loss: 0.2940 - val_mse: 0.0871
Epoch 4/4
782/782 [==============================] - 3s 4ms/step - loss: 0.2941 - mse: 0.0879
[0.2941144108772278, 0.08792287856340408]


The loss function change from Binary crossentropy to MSE has decreased the accuracy from 85% to 0.08%